File: src/tizenclaw/anthropic_backend.cc
#include "anthropic_backend.hh"
#include "http_client.hh"

#include "../common/logging.hh"

namespace tizenclaw {


bool AnthropicBackend::Initialize(
    const nlohmann::json& config) {
  api_key_ = config.value("api_key", "");
  model_ = config.value("model",
      "claude-sonnet-4-20250514");
  if (api_key_.empty()) {
    LOG(ERROR) << "Anthropic API key is empty";
    return false;
  }
  LOG(INFO) << "Anthropic backend initialized (model: " << model_ << ")";
  return true;
}

nlohmann::json AnthropicBackend::ToAnthropicMessages(
    const std::vector<LlmMessage>& messages) const {
  nlohmann::json msgs = nlohmann::json::array();
  for (auto& msg : messages) {
    nlohmann::json entry;

    if (msg.role == "user") {
      entry = {{"role", "user"},
               {"content", msg.text}};
    } else if (msg.role == "assistant") {
      entry["role"] = "assistant";
      if (!msg.tool_calls.empty()) {
        nlohmann::json content =
            nlohmann::json::array();
        for (auto& tc : msg.tool_calls) {
          content.push_back({
              {"type", "tool_use"},
              {"id", tc.id},
              {"name", tc.name},
              {"input", tc.args}
          });
        }
        entry["content"] = content;
      } else {
        entry["content"] = msg.text;
      }
    } else if (msg.role == "tool") {
      entry = {
          {"role", "user"},
          {"content", {{
              {"type", "tool_result"},
              {"tool_use_id", msg.tool_call_id},
              {"content",
               msg.tool_result.dump()}
          }}}
      };
    }
    msgs.push_back(entry);
  }
  return msgs;
}

nlohmann::json AnthropicBackend::ToAnthropicTools(
    const std::vector<LlmToolDecl>& tools) const {
  if (tools.empty()) return nullptr;
  nlohmann::json result = nlohmann::json::array();
  for (auto& t : tools) {
    result.push_back({
        {"name", t.name},
        {"description", t.description},
        {"input_schema", t.parameters}
    });
  }
  return result;
}

LlmResponse
AnthropicBackend::ParseAnthropicResponse(
    const std::string& body) const {
  LlmResponse resp;
  try {
    auto j = nlohmann::json::parse(body);

    if (j.contains("error")) {
      resp.success = false;
      resp.error_message =
          j["error"].value("message",
                           "Unknown error");
      return resp;
    }

    if (!j.contains("content") ||
        j["content"].empty()) {
      resp.success = false;
      resp.error_message = "Empty content";
      return resp;
    }

    resp.success = true;
    for (auto& block : j["content"]) {
      std::string type =
          block.value("type", "");
      if (type == "text") {
        if (!resp.text.empty())
          resp.text += "\n";
        resp.text +=
            block["text"].get<std::string>();
      } else if (type == "tool_use") {
        LlmToolCall tc;
        tc.id = block.value("id", "");
        tc.name = block["name"];
        tc.args = block["input"];
        resp.tool_calls.push_back(tc);
      }
    }
  } catch (const std::exception& e) {
    resp.success = false;
    resp.error_message =
        std::string("Parse error: ") + e.what();
  }
  return resp;
}

LlmResponse AnthropicBackend::Chat(
    const std::vector<LlmMessage>& messages,
    const std::vector<LlmToolDecl>& tools,
    std::function<void(const std::string&)> on_chunk) {
  nlohmann::json payload = {
      {"model", model_},
      {"max_tokens", 4096},
      {"messages",
       ToAnthropicMessages(messages)}
  };
  auto ant_tools = ToAnthropicTools(tools);
  if (!ant_tools.is_null()) {
    payload["tools"] = ant_tools;
  }

  std::string url =
      "https://api.anthropic.com/v1/messages";

  auto http_resp = HttpClient::Post(
      url,
      {{"Content-Type", "application/json"},
       {"x-api-key", api_key_},
       {"anthropic-version", "2023-06-01"}},
      payload.dump(),
      3, 10, 30, on_chunk); // Pass on_chunk callback

  if (!http_resp.success) {
    LlmResponse r;
    r.success = false;
    r.error_message = http_resp.error;
    if (!http_resp.body.empty()) {
      try {
        auto ej =
            nlohmann::json::parse(http_resp.body);
        if (ej.contains("error")) {
          r.error_message += ": " +
              ej["error"].value("message", "");
        }
      } catch (...) {
        r.error_message += ": " +
            http_resp.body.substr(
                0, std::min((size_t)200,
                            http_resp.body.size()));
      }
    }
    LOG(ERROR) << "API error: "
               << r.error_message;
    return r;
  }

  return ParseAnthropicResponse(http_resp.body);
}

} // namespace tizenclaw

File: src_py/tizenclaw/llm/anthropic_backend.py
"""
Anthropic (Claude) LLM Backend for TizenClaw.

Uses the Anthropic Messages API (api.anthropic.com).
Handles Anthropic's unique header requirements and message format.
"""
import asyncio
import json
import logging
import os
import ssl
import urllib.request
from typing import Dict, List, Any, Optional, Callable

from .llm_backend import (
    LlmBackend, LlmMessage, LlmToolDecl, LlmToolCall, LlmResponse
)

logger = logging.getLogger(__name__)


class AnthropicBackend(LlmBackend):
    """Anthropic Claude API backend."""

    def __init__(self):
        self._api_key = ""
        self._model = "claude-sonnet-4-20250514"
        self._endpoint = "https://api.anthropic.com/v1/messages"
        self._ssl_ctx: Optional[ssl.SSLContext] = None
        self._backend_name = "anthropic"
        self._api_version = "2023-06-01"

    def get_name(self) -> str:
        return f"anthropic/{self._model}"

    async def initialize(self, config: Dict[str, Any]) -> bool:
        self._api_key = config.get("api_key", os.environ.get("ANTHROPIC_API_KEY", ""))
        self._model = config.get("model", self._model)
        ep = config.get("endpoint", "")
        if ep:
            self._endpoint = ep.rstrip("/")

        if not self._api_key:
            logger.error("Anthropic: no API key provided")
            return False

        self._ssl_ctx = ssl.create_default_context()
        for ca in ["/etc/ssl/ca-bundle.pem", "/etc/ssl/certs/ca-certificates.crt",
                    "/usr/share/ca-certificates"]:
            if os.path.exists(ca):
                try:
                    if os.path.isfile(ca):
                        self._ssl_ctx.load_verify_locations(ca)
                    else:
                        self._ssl_ctx.load_verify_locations(capath=ca)
                    break
                except Exception:
                    pass

        logger.info(f"Initialized {self.get_name()} → {self._endpoint}")
        return True

    def _convert_messages(self, messages: List[LlmMessage]) -> List[Dict]:
        """Convert to Anthropic message format."""
        result = []
        for m in messages:
            if m.role == "system":
                continue  # System prompt handled separately

            if m.role == "assistant":
                content = []
                if m.content:
                    content.append({"type": "text", "text": m.content})
                if m.tool_calls:
                    for tc in m.tool_calls:
                        content.append({
                            "type": "tool_use",
                            "id": tc.id or tc.name,
                            "name": tc.name,
                            "input": tc.args if isinstance(tc.args, dict) else {}
                        })
                result.append({"role": "assistant", "content": content})

            elif m.role == "tool":
                result.append({
                    "role": "user",
                    "content": [{
                        "type": "tool_result",
                        "tool_use_id": m.tool_call_id or "",
                        "content": m.content or ""
                    }]
                })

            else:  # user
                result.append({"role": "user", "content": m.content or ""})

        return result

    def _convert_tools(self, tools: List[LlmToolDecl]) -> List[Dict]:
        """Convert to Anthropic tool format."""
        result = []
        for t in tools:
            schema = t.parameters or {
                "type": "object",
                "properties": {
                    "arguments": {
                        "type": "string",
                        "description": "Arguments for the tool"
                    }
                }
            }
            result.append({
                "name": t.name,
                "description": t.description or "",
                "input_schema": schema
            })
        return result

    def _make_request(self, payload: Dict) -> Dict:
        body = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(
            self._endpoint, data=body, method="POST",
            headers={
                "Content-Type": "application/json",
                "x-api-key": self._api_key,
                "anthropic-version": self._api_version,
            }
        )

        try:
            resp = urllib.request.urlopen(req, context=self._ssl_ctx, timeout=120)
            return json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            error_body = e.read().decode("utf-8", errors="replace")
            logger.error(f"Anthropic API error {e.code}: {error_body[:200]}")
            return {"error": f"HTTP {e.code}: {error_body[:200]}", "http_status": e.code}
        except Exception as e:
            logger.error(f"Anthropic request failed: {e}")
            return {"error": str(e)}

    async def chat(
        self,
        messages: List[LlmMessage],
        tools: List[LlmToolDecl],
        on_chunk: Optional[Callable[[str], None]] = None,
        system_prompt: str = ""
    ) -> LlmResponse:
        api_messages = self._convert_messages(messages)

        payload: Dict[str, Any] = {
            "model": self._model,
            "messages": api_messages,
            "max_tokens": 4096,
        }

        if system_prompt:
            payload["system"] = system_prompt

        converted_tools = self._convert_tools(tools)
        if converted_tools:
            payload["tools"] = converted_tools

        logger.info(f"Anthropic API call: {self._model}, messages={len(api_messages)}, tools={len(converted_tools)}")

        result = await asyncio.to_thread(self._make_request, payload)

        if "error" in result:
            err_msg = result["error"]
            if isinstance(err_msg, dict):
                err_msg = err_msg.get("message", str(err_msg))
            return LlmResponse(
                success=False,
                error_message=str(err_msg),
                http_status=result.get("http_status", 0)
            )

        # Parse response
        content_blocks = result.get("content", [])
        text = ""
        tcalls = []

        for block in content_blocks:
            if block.get("type") == "text":
                text += block.get("text", "")
            elif block.get("type") == "tool_use":
                tcalls.append(LlmToolCall(
                    id=block.get("id", ""),
                    name=block.get("name", ""),
                    args=block.get("input", {})
                ))

        usage = result.get("usage", {})
        logger.info(f"Anthropic response: text={len(text)} chars, tool_calls={len(tcalls)}, tokens={usage.get('output_tokens', 0)}")

        return LlmResponse(
            success=True,
            text=text,
            tool_calls=tcalls,
            prompt_tokens=usage.get("input_tokens", 0),
            completion_tokens=usage.get("output_tokens", 0),
            total_tokens=usage.get("input_tokens", 0) + usage.get("output_tokens", 0)
        )

    async def generate_stream(self, prompt, history, tools):
        resp = await self.chat(
            history + [LlmMessage(role="user", content=prompt)],
            tools
        )
        return resp
